subprojects/core/src/main/java/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistry.java

```diff
@@ -19,26 +19,60 @@
 import com.google.common.collect.Sets;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.file.FileResolver;
-import org.gradle.api.internal.file.UnionFileCollection;
 
+import java.io.File;
+import java.nio.file.Path;
+import java.util.LinkedHashSet;
 import java.util.Set;
 
+@SuppressWarnings("Since15")
 public class DefaultBuildOutputCleanupRegistry implements BuildOutputCleanupRegistry {
 
     private final FileResolver fileResolver;
     private final Set<FileCollection> outputs = Sets.newHashSet();
+    private Set<Path> resolvedFiles;
 
     public DefaultBuildOutputCleanupRegistry(FileResolver fileResolver) {
         this.fileResolver = fileResolver;
     }
 
     @Override
-    public void registerOutputs(Object files) {
+    public synchronized void registerOutputs(Object files) {
+        if (resolvedFiles != null) {
+            resolvedFiles = null;
+        }
         this.outputs.add(fileResolver.resolveFiles(files));
     }
 
     @Override
-    public FileCollection getOutputs() {
-        return new UnionFileCollection(outputs);
+    public boolean isSaveToDelete(File file) {
+        Set<Path> safeToDelete = getResolvedPaths();
+        Path absolutePath = file.toPath().toAbsolutePath();
+        do {
+            if (safeToDelete.contains(absolutePath)) {
+                return true;
+            }
+            absolutePath = absolutePath.getParent();
+        } while (absolutePath != null);
+        return false;
+    }
+
+    private Set<Path> getResolvedPaths() {
+        if (resolvedFiles == null) {
+            doResolvePaths();
+        }
+        return resolvedFiles;
+    }
+
+    private synchronized void doResolvePaths() {
+        if (resolvedFiles == null) {
+            Set<Path> set = new LinkedHashSet<Path>();
+            for (FileCollection output : outputs) {
+                for (File file : output.getFiles()) {
+                    set.add(file.toPath().toAbsolutePath());
+                }
+            }
+            resolvedFiles = set;
```

Comment:
Should we call this `resolvedPaths`? Also, `set` is not a cool name for a variable. :)